loom/engine/bindings/loom/lmSQLite.cpp

```diff
@@ -0,0 +1,647 @@
+    }
+    return s;
+}
+
+Connection *Connection::open(const char *database, int flags)
+{
+    Connection *c;
+    int res;
+
+    //check for valid database name
+    if((database == NULL) || (database[0] == '\0'))
+    {
+        lmLogError(gSQLiteGroup, "Invalid database name specified!");
+        return NULL;
+    }
+
+    //create the connection
+    c = new Connection();
+    c->databaseName = utString(database);
+
+    //if we find a path separator in the database name, we assume that it contains a valid path already
+    if(strchr(database, '/') || strchr(database, '\\'))
+    {
+        c->databaseFullPath = c->databaseName;
+    }
+    else
+    {
+        //no separator, so we need to prefix the system writable path in front of the database name
+        c->databaseFullPath = utString(cocos2d::CCFileUtils::sharedFileUtils()->getWriteablePath().c_str()) + c->databaseName;
+    }
+    
+    //open the SQLite DB
+    res = sqlite3_open_v2(c->databaseFullPath.c_str(), &c->dbHandle, flags, 0);
+    if(res != SQLITE_OK)
+    {
+        lmLogError(gSQLiteGroup, "Error opening the SQLite database file: %s with message: %s", c->databaseFullPath.c_str(), c->getErrorMessage());
+    }
+    return c;
+}
+
+bool Connection::backgroundImport(const char *database, JSON *data)
+{
+    if(Connection::backgroundImportInProgress)
+    {
+        lmLogError(gSQLiteGroup, "Attempting to run multiple backgroundImport calls on the same database: %s", database);
+        return false;
+    }
+
+    //store values to be used in the thread
+    Connection::backgroundImportDatabase = database;
+    Connection::backgroundImportData = data;
+
+    //start up the backgroundImport thread
+    Connection::backgroundImportInProgress = true;
+    loom_thread_start(Connection::backgroundImportBody, NULL);    
+    return true;
+}
+
+const char* Connection::getVersion()
+{
+    return sqlite3_libversion();
+}
+
+void Connection::backgroundImportDone(int result)
+{    
+    //fire our completion delegate
+    Connection::_OnImportCompleteDelegate.pushArgument(result);
+    Connection::_OnImportCompleteDelegate.invoke();
+
+    //we're done now, so need to turn off our processing flag
+    loom_mutex_lock(Connection::backgroundImportMutex);
+    Connection::backgroundImportInProgress = false;
+    loom_mutex_unlock(Connection::backgroundImportMutex);
+}
+
+int __stdcall Connection::backgroundImportBody(void *param)
+{
+    int i, j;
+    int numRows;
+    JSON *data;
+    JSON *table;
+    JSON *row;
+    Connection *c;
+    Statement *s;
+    utString query;
+    const char *database;
+    const char *tableName;
+    const char *columnName;
+    const char *firstColumn;
+
+
+    //get the database name and JSON data
+    database = Connection::backgroundImportDatabase;
+    data = Connection::backgroundImportData;
+
+///////////
+///////////
+//BEN QUESTION: Do we need to handle CREATE of a database as well? 
```

Comment:
Cool. As it is, it should fail with relevant error messages but I'll force
to double check.
On 17 Feb 2015 3:30 AM, "Ben Garney" <notifications@github.com> wrote:

> In loom/engine/bindings/loom/lmSQLite.cpp
> <https://github.com/LoomSDK/LoomSDK/pull/261#discussion_r24786973>:
>
> > +    Connection *c;
> > +    Statement *s;
> > +    utString query;
> > +    const char *database;
> > +    const char *tableName;
> > +    const char *columnName;
> > +    const char *firstColumn;
> > +
> > +
> > +    //get the database name and JSON data
> > +    database = Connection::backgroundImportDatabase;
> > +    data = Connection::backgroundImportData;
> > +
> > +///////////
> > +///////////
> > +//BEN QUESTION: Do we need to handle CREATE of a database as well?
>
> We can assume DB is created... Can even assume tables are initialized. But
> should have good error handling for either assumption being wrong.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/LoomSDK/LoomSDK/pull/261/files#r24786973>.
>